### PR TITLE
sql: fix a panic in TestSQLEncode

### LINKED ISSUE #29545: sql: panic in lex.TestEncodeSQLBytes when given invalid inputs
Found via

```
 make testlogic TESTFLAGS='-verbosity=10 -show-sql -test.v' FILES=cluster_version
...
SET CLUSTER SETTING version = '1.0-500';
panic: runtime error: slice bounds out of range

goroutine 418 [running]:
github.com/cockroachdb/cockroach/pkg/sql/lex.EncodeSQLStringWithFlags(0xc420f59588, 0xc421498520, 0x16, 0x0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/lex/encode.go:112 +0x3a3
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DString).Format(0xc42162f6a0, 0xc420f59560)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1120 +0x8b
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrHideConstants(0xc420f59560, 0x2787980, 0xc42162f6a0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go:47 +0xa4
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0xc420f59560, 0x2787980, 0xc42162f6a0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:256 +0x383
github.com/cockroachdb/cockroach/pkg/sql.(*explainer).expr(0xc4212c6f80, 0x22c660f, 0x6, 0xc42041cca0, 0xb, 0x1, 0x27af640, 0xc42162f6a0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:364 +0x129
github.com/cockroachdb/cockroach/pkg/sql.(*explainer).(github.com/cockroachdb/cockroach/pkg/sql.expr)-fm(0x22c660f, 0x6, 0xc42041cca0, 0xb, 0x1, 0x27af640, 0xc42162f6a0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:344 +0x73
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).expr(0xc4211da460, 0x22c660f, 0x6, 0xc42041cca0, 0xb, 0x1, 0x27af640, 0xc42162f6a0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:560 +0x8a
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc4211da460, 0x27a3080, 0xc4211da050, 0x22c660f, 0x6)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:161 +0x24e
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc4211da460, 0x27a3080, 0xc4211da050, 0x22c77f3, 0x7)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:94 +0x8f
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc4211da460, 0x27a3040, 0xc421050690, 0x22c65bb, 0x6)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:439 +0x2d42
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitConcrete(0xc4211da460, 0x27a3040, 0xc421050690)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:107 +0x83
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc4211da460, 0x27a2a00, 0xc42039cd20, 0x22c3fe6, 0x5)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:477 +0x396d
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc4211da460, 0x27a2a00, 0xc42039cd20, 0x10, 0xc42104dd90)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:94 +0x8f
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(0x27acf40, 0xc4215aa7e0, 0x27a2a00, 0xc42039cd20, 0x0, 0xc42104dd60, 0xc42104dd70, 0xc42104dd80, 0xc42104dd90, 0x6, ...)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:58 +0x193
github.com/cockroachdb/cockroach/pkg/sql.(*explainer).populateEntries(0xc4212c6f80, 0x27acf40, 0xc4215aa7e0, 0x27a2a00, 0xc42039cd20, 0x0, 0x0, 0x0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:290 +0x1eb
github.com/cockroachdb/cockroach/pkg/sql.planToString(0x27acf40, 0xc4215aa7e0, 0x27a2a00, 0xc42039cd20, 0x0, 0x0, 0x0, 0x0, 0x27a2a00)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:320 +0x150
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makePlan(0xc4211a8a38, 0x27acf40, 0xc4215aa7e0, 0x27afcc0, 0xc4214a48c0, 0x3877798, 0x0, 0x0, 0xc4205b9b60, 0x5f, ...)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:350 +0x5b4
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc4211a8600, 0x27acf40, 0xc4215aa7e0, 0x27afcc0, 0xc4214a48c0, 0x3877798, 0x0, 0x0, 0xc4205b9b60, 0x5f, ...)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:756 +0xdc6
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc4211a8600, 0x27acf40, 0xc4215aa7e0, 0x27afcc0, 0xc4214a48c0, 0x3877798, 0x0, 0x0, 0xc4205b9b60, 0x5f, ...)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:401 +0xb82
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc4211a8600, 0x27acf40, 0xc4215aa7e0, 0x27afcc0, 0xc4214a48c0, 0x3877798, 0x0, 0x0, 0xc4205b9b60, 0x5f, ...)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:95 +0x358
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc4211a8600, 0x27acf40, 0xc4214aa900, 0x0, 0x0, 0x0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1135 +0x13cd
github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).initConnEx.func1(0xc4211a8600, 0x27acf40, 0xc4214aa900, 0xc420421a20, 0xc4216a0f00, 0xc42138dfc0)
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:210 +0x8a
created by github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).initConnEx
	/home/arjun/goworkspace/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:199 +0x275
Makefile:864: recipe for target 'testbaselogic' failed
make: *** [testbaselogic] Error 2

```

Upon further inspection, `TestEncodeSQLBytes` is not resilient to malformed string inputs.

### LINKED ISSUE #29636: sql: select substr(b'\275', 0, 1) panics
The stack trace is... very long, but I think what matters is this:

```
panic(0x61a4e00, 0x7f49260)
        /usr/local/Cellar/go/1.10.3/libexec/src/runtime/panic.go:502 +0x229
github.com/cockroachdb/cockroach/pkg/util/stringencoding.EncodeEscapedChar(0xc42145da98, 0xc420396d28, 0x1, 0xfd0000fffd, 0x0, 0x27)
        /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/util/stringencoding/string_encoding.go:142 +0x3c4
github.com/cockroachdb/cockroach/pkg/sql/lex.EncodeSQLStringWithFlags(0xc42145da98, 0xc420396d28, 0x1, 0x0)
        /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/sql/lex/encode.go:119 +0x139
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DString).Format(0xc421424698, 0xc42145da70)
        /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1120 +0x8b
```

cc @jordanlewis

### PR DESCRIPTION
**Pull request by rjnn**
Fixes #29545. Fixes #29636.

This change involves additional test cases that are CC-by-4.0 licensed.

### CODE REVIEW

**Review by jordanlewis — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29549)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/lex/encode_test.go, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29549#-LLa16LQEzBLN2Yoodg-:-LLa16LQEzBLN2Yoodg0:b-2kbxdy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0ad2ef1626b83f7d7266436a53271f1f729742/pkg/sql/lex/encode_test.go#L90)):*
> ```Go
> 		"                                                                              |",
> 		"3.1.3  2 continuation bytes: \"��\"                                             |",
> 		"3.1.4  3 continuation bytes: \"���\"                                            |",
> ```

I hate to ask, but are these really testing what we want if we embed them in longer strings? Wouldn't we have to isolate all of the strings to make sure we're testing en/decoding the actual utf8 strings themselves? To make this a really good test I think we'd also have to make expected output... though I do understand this is a pain and we're just trying to catch egregious crash bugs right now. Maybe it's fine as-is. Curious what @knz thinks.

**Comment by arjunravinarayan** [comment]
___
*[pkg/sql/lex/encode_test.go, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29549#-LLa16LQEzBLN2Yoodg-:-LLeGy4d4to2YBCtLg6N:baygqpp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0ad2ef1626b83f7d7266436a53271f1f729742/pkg/sql/lex/encode_test.go#L90)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I hate to ask, but are these really testing what we want if we embed them in longer strings? Wouldn't we have to isolate all of the strings to make sure we're testing en/decoding the actual utf8 strings themselves? To make this a really good test I think we'd also have to make expected output... though I do understand this is a pain and we're just trying to catch egregious crash bugs right now. Maybe it's fine as-is. Curious what @knz thinks.
</blockquote></details>

These are the test files as-is, and I think there's some value to them. Yes, we do want end-of-string tests, and that's why I added that one-off known-to-previously-crash test case at the top.

We don't have a decoder, so I can't do a roundtrip test. However, I guess I could find a known-good UTF-8 encoder, run these inputs through them, and check that we match that output exactly.

The primary motivation of this PR is to fix a panic, shoring up the testing is incidental. Would you mind if I filed an issue and moved on?

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29549)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/lex/encode_test.go, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29549#-LLa16LQEzBLN2Yoodg-:-LLeNhpk07RpF3qXfW6g:b-885qkd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0ad2ef1626b83f7d7266436a53271f1f729742/pkg/sql/lex/encode_test.go#L90)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

These are the test files as-is, and I think there's some value to them. Yes, we do want end-of-string tests, and that's why I added that one-off known-to-previously-crash test case at the top.

We don't have a decoder, so I can't do a roundtrip test. However, I guess I could find a known-good UTF-8 encoder, run these inputs through them, and check that we match that output exactly.

The primary motivation of this PR is to fix a panic, shoring up the testing is incidental. Would you mind if I filed an issue and moved on?
</blockquote></details>

Fair enough. I'd say don't bother filing the issue if you think this is fine as well - I just couldn't help mentioning the obvious :)

**Review by jordanlewis — APPROVED** [review]

**Review by rjnn — COMMENTED** [review]
TFTR @jordanlewis. I'm going to hold this PR until @bdarnell is back and OKs the license addition.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29549#-:-LLehk598uUOodeeiHKb:bjqun2f)* status: :shipit: complete! 0 of 0 LGTMs obtained

File: pkg/sql/lex/encode_test.go [hunk 1]
```diff
@@ -34,6 +34,263 @@ func TestEncodeSQLString(t *testing.T) {
 	testEncodeSQL(t, lex.EncodeSQLString, true)
 }
 
+func TestEncodeSQLString2(t *testing.T) {
+	testCases := []string{
+		// Regression test for #
```
Comment (tschottdorf) [inline pkg/sql/lex/encode_test.go]: missing issue number

**Review by rjnn — COMMENTED** [review]
I have removed the CC-by-4.0 test cases (and thus the license gating). That can wait for later. PTAL @jordanlewis. I now have a simple logic test that ensures that we error out correctly when we encounter invalid UTF-8 byte strings. This is tested only with DistSQL, as local SQL does not serialize and deserialize the string, it simply pretty prints out the escaped invalid UTF-8 as-is. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29549#-:-LLkKv38B9z_gWmZIbZQ:b-3zy234)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/lex/encode_test.go, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/29549#-LLa0QSldHN1PUXXmScy-r2-39:-LLkLC049qiVMsqVAxtb:b-jrlzrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/186133b54c9a95f1dec25128e1012c982d67b823/pkg/sql/lex/encode_test.go#L39)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

missing issue number
</blockquote></details>

No longer relevant, I've decided not to add tests here. Thank you.

**Review by jordanlewis — APPROVED** [review]
LGTM thanks!

**Comment by arjunravinarayan** [comment]
bors r=jordanlewis

**Pull request merged**